src/main/java/com/redhat/insights/agent/AgentSubreport.java

```diff
@@ -124,47 +124,55 @@ static String fingerprintJBoss(Class<?> moduleClass) {
-      Object moduleLoader = getBootModuleLoaderMethod.invoke(null, EMPTY_OBJECT_ARRAY);
-      ClassLoader versionModuleClassLoader =
-          getModuleClassLoader(moduleLoader, "org.jboss.as.version");
-      String home = getJBossHome();
-      Class<?> moduleLoaderClass = getJBossModuleLoaderClass(moduleLoader.getClass());
-      Class<?> productConfigClass =
-          versionModuleClassLoader.loadClass("org.jboss.as.version.ProductConfig");
-      Method fromFilesystemSlotMethod =
-          productConfigClass.getDeclaredMethod(
-              "fromFilesystemSlot", moduleLoaderClass, String.class, Map.class);
-      Object productConfig =
-          fromFilesystemSlotMethod.invoke(
-              null, moduleLoaderClass.cast(moduleLoader), home, getPropertiesPrivileged());
-      return (String)
-          productConfigClass
-              .getDeclaredMethod("getPrettyVersionString", EMPTY_CLASS_ARRAY)
-              .invoke(productConfig, EMPTY_OBJECT_ARRAY);
-
-    } catch (NoSuchMethodException
-        | SecurityException
-        | ClassNotFoundException
-        | IllegalAccessException
-        | IllegalArgumentException
-        | InvocationTargetException ex) {
-      // ignore
+      // Wait until JBoss Modules initializes to avoid breaking their module loader
+      String modulePath = getJBossModulePath();
+      if (modulePath == null) {
+        logger.debug("Module path did not become available");
+      } else {
+        Object moduleLoader = getBootModuleLoaderMethod.invoke(null, EMPTY_OBJECT_ARRAY);
+        ClassLoader versionModuleClassLoader =
+            getModuleClassLoader(moduleLoader, "org.jboss.as.version");
+        String home = getJBossHome();
+        Class<?> moduleLoaderClass = getJBossModuleLoaderClass(moduleLoader.getClass());
+        Class<?> productConfigClass =
+            versionModuleClassLoader.loadClass("org.jboss.as.version.ProductConfig");
+        Method fromFilesystemSlotMethod =
+            productConfigClass.getDeclaredMethod(
+                "fromFilesystemSlot", moduleLoaderClass, String.class, Map.class);
+        Object productConfig =
+            fromFilesystemSlotMethod.invoke(
+                null, moduleLoaderClass.cast(moduleLoader), home, getPropertiesPrivileged());
+        return (String)
+            productConfigClass
+                .getDeclaredMethod("getPrettyVersionString", EMPTY_CLASS_ARRAY)
+                .invoke(productConfig, EMPTY_OBJECT_ARRAY);
+      }
+    } catch (Exception ex) {
+      logger.debug("Ignoring exception during JBoss probe", ex);
     }
     return "Unknown EAP / Wildfly - possibly misconfigured";
   }
 
   private static String getJBossHome() {
+    return pollProperty("jboss.home.dir");
+  }
+
+  private static String getJBossModulePath() {
+    return pollProperty("module.path");
+  }
+
+  private static String pollProperty(String property) {
     long timeout = System.currentTimeMillis() + 3000L;
```

Comment:
I'm not convinced this timeout is long enough - can we make it 10s and pull it out into a proper constant?

Reply:
Done